content/blog/2022-03-08-alive/index.md

```diff
@@ -0,0 +1,114 @@
+
+### Usability
+In my opinion, the usability and real-world usage of the Alive toolchain is itself a notable contribution. It would be one thing if the authors used a fancy method to find some bugs that they reported and got fixed, but it’s a much more impressive thing if other people can use the author’s work on their own problems. 
+
+#### Toolchain
+The authors were able to build a project that other people both wanted to use and contribute to. This project still has users![^2]
+
+[^2]: Technically, Alive 2 is still maintained, which uses SMT solvers for translation validation of LLVM's test suite. 
+
+#### Know your audience
+The authors selected an audience (LLVM contributors) who they knew would be interested in using the tool to make their lives easier. This was enough for some contributors to be using the tool even before the paper was published. The authors were also able to show users the value of their tool by monitoring proposed patches to LLVM, finding bugs in them, and working with contributors to fix them. This greatly improved the impact of the work by getting LLVM contributors to use the tool to continue fixing and avoiding bugs even after the paper was published.
+
+## Discussion
+While many of the great points from class have already been discussed, there are a few other points that didn’t really make it into the blog post so far.
+
+### Limitations
+The paper has a few limitations which the authors also mention, such as not being able to translate all of InstCombine because it doesn't handle vectors, call/return instructions, or floating point operation.
+
+More deeply, its reliance on an SMT solver means that as more variables, operations, or bits get added to an optimization, the harder it is to verify the optimization, without there really being any way to fix the method. The paper already has issues with the handling larger bitwidths quickly for multiplication and division, and it seems that it would be hard to scale this method to handle larger bitwidths or expressions with more variables.
+
+### Defining Undefined for Fun and Profit
+One possibly-confusing thing in the paper is that the SMT solver can choose any value that it finds convenient for any `undef` value. While a glib justification may simply be that LLVM and C say that anything can happen, I think this is actually a quite reasonable and good thing for Alive to do. By choosing convenient values for `undef`, it maximizes its ability to verify specifications which agree on the values that are correct. It would have been interesting for the paper to report on how many of the successful verifications depended on this behavior.
```

Comment:
Not that you have to change anything here, but TBH I personally don't see any _other_ good way to deal with `undef` when modeling its semantics in SMT. Just picking zero deterministically would at least be correct, for example, but it doesn't seem to obey the spirit of verifying optimizations in the presence of `undef`: that is, it doesn't seem to help prove optimizations incorrect when they do something wrong with `undef`.